### ISSUE TITLE
m68k clone.S contains unsupported instructions for m68000 cpu

### ISSUE THREAD

**Issue by GiovanniScotti** [wbx-github/uclibc-ng#15]
Hi all,
I noticed that the clone.S provided for compatibility with m68k architecture contains unsupported instructions by m68000.
In particular, I noticed that "tstl" instruction is used with address registers. Despite the fact this is supported by 68020 and newer, a toolchain built for 68000 will complain.

The following is part of git commit 23bb649090ff588e8642f0c581cfe7ce2d29c757 for clone.S and shows where the problem has been introduced:

```
 +__clone:
        /* Sanity check arguments.  */
-       movel   #-EINVAL, %d0
-       movel   4(%sp), %d1             /* no NULL function pointers */
-       movel   %d1, %a0
-       tstl    %d1
+       movel   #-EINVAL, %d0
+       movel   4(%sp), %a0             /* no NULL function pointers */
+       tstl    %a0
        beq.w   __syscall_error_trampoline
-       movel   8(%sp), %d1             /* no NULL stack pointers */
-       movel   %d1, %a1
-       tstl    %d1
+       movel   8(%sp), %a1             /* no NULL stack pointers */
+       tstl    %a1
```
As you can see, movel+tstl have been replaced by tstl directly using an address register.

My use case is using a custom Buildroot checkout that supports 68000 (no MMU, FLAT binary) used to generate a cross toolchain for 68000 cpu. During toolchain compilation, the process stops due to the issue above.
Thank you for your support.

**Comment by wbx-github** [comment]
Can you sent a patch reverting the change with your SoB to the uclibc-ng mailinglist? What hardware do you use? The rest is working?

**Comment by GiovanniScotti** [comment]
Hi, thank you for your reply.
My goal is running the latest Linux kernel (6.9) on a very simple home made 68000-based architecture providing 8MB of RAM (don't ask why, just for having fun in my spare time). The system has a simple bootloader that loads a binary image from a compact flash (CF card) to a known address in RAM and jumps there. Since 68000 has not a MMU, linux is compiled as ucLinux.

I need to and would like to use a modern toolchain to do that, but I'm facing several issues building a m68k-uclinux-uclibc- one. It seems that at some point in time (around version 4.2.4), gcc has introduced some incompatibility with the 68000 cpu. In particular, the cross compiler generates unaligned memory accesses because assumes a 68020 cpu even if option -m68000 is provided at compile time. Have a look here:

https://stackoverflow.com/questions/74956839/gcc-produces-code-with-unaligned-address-error-on-m68k-when-defining-a-local-str

There is also a bug opened on gcc Bugzilla (never fixed), where a guy proposed a workaround to make it work:

https://gcc.gnu.org/bugzilla/show_bug.cgi?id=53386

By applying the patch at "comment 1", I successfully compiled a working gcc 13.2 that allows me to compile Linux for 68000 and boot it without exceptions. My system is now able to boot the kernel up to "Run /init as init process", but I have not built a root filesystem yet. Anyway, during the compilation of the cross toolchain using Buildroot, I encountered the following error:

  AS libc/sysdeps/linux/m68k/clone.o
libc/sysdeps/linux/m68k/clone.S: Assembler messages:
libc/sysdeps/linux/m68k/clone.S:40: Error: operands mismatch -- statement \`tstl %a0' ignored
libc/sysdeps/linux/m68k/clone.S:43: Error: operands mismatch -- statement \`tstl %a1' ignored

The error is the one I explained in my previous message. It goes away patching the clone.S in this way:

```
__clone:
   [ ... ]
   movel   4(%sp), %d1             /* no NULL function pointers */
   movel   %d1, %a0
   tstl    %d1
   [ ... ]
   movel   8(%sp), %d1             /* no NULL stack pointers */
   movel   %d1, %a1
   tstl    %d1
```

I plan to create an initramfs cpio archive by using busybox or buildroot itself and embed it in the kernel image, but I haven't succeded yet (I tried, but the cpu detects exceptions). Thus, I'm planning to create a very very simple init flat binary, compile it, build cpio, embed it into kernel, load it and see what happens. For the sake of clarity, I will compile in such way:

`m68k-nommu-uclinuc-uclibc-gcc -fPIC -Wl,elf2flt=-r -Wall -static -o init init.c`

I can tell you that "the rest is working" from a compile-time point of view, but not from a run-time one, because I'm still not able to run a user space application.
